<a href="https://colab.research.google.com/github/Jon227student/IS_170_Machine_Learning/blob/7.-K-Nearest-Neighbors/HW_7.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
import pandas as pd
import numpy as np
from sklearn import preprocessing 

In [ ]:
df=pd.read_csv('/content/Loan_Default.csv')

In [ ]:
missing_values = df.isnull().sum()
print(missing_values)

ID                               0
year                             0
loan_limit                    3344
Gender                           0
approv_in_adv                  908
loan_type                        0
loan_purpose                   134
Credit_Worthiness                0
open_credit                      0
business_or_commercial           0
loan_amount                      0
rate_of_interest             36439
Interest_rate_spread         36639
Upfront_charges              39642
term                            41
Neg_ammortization              121
interest_only                    0
lump_sum_payment                 0
property_value               15098
construction_type                0
occupancy_type                   0
Secured_by                       0
total_units                      0
income                        9150
credit_type                      0
Credit_Score                     0
co-applicant_credit_type         0
age                            200
submission_of_applic

In [ ]:
df = df.drop(['Gender', 'year'], axis=1)
print(df)

            ID loan_limit approv_in_adv loan_type loan_purpose  \
0        24890         cf         nopre     type1           p1   
1        24891         cf         nopre     type2           p1   
2        24892         cf           pre     type1           p1   
3        24893         cf         nopre     type1           p4   
4        24894         cf           pre     type1           p1   
...        ...        ...           ...       ...          ...   
148665  173555         cf         nopre     type1           p3   
148666  173556         cf         nopre     type1           p1   
148667  173557         cf         nopre     type1           p4   
148668  173558         cf         nopre     type1           p4   
148669  173559         cf         nopre     type1           p3   

       Credit_Worthiness open_credit business_or_commercial  loan_amount  \
0                     l1        nopc                  nob/c       116500   
1                     l1        nopc                   

In [ ]:
df = df.drop(['ID'], axis=1)

In [ ]:
print(df.head())

  loan_limit approv_in_adv loan_type loan_purpose Credit_Worthiness  \
0         cf         nopre     type1           p1                l1   
1         cf         nopre     type2           p1                l1   
2         cf           pre     type1           p1                l1   
3         cf         nopre     type1           p4                l1   
4         cf           pre     type1           p1                l1   

  open_credit business_or_commercial  loan_amount  rate_of_interest  \
0        nopc                  nob/c       116500               NaN   
1        nopc                    b/c       206500               NaN   
2        nopc                  nob/c       406500              4.56   
3        nopc                  nob/c       456500              4.25   
4        nopc                  nob/c       696500              4.00   

   Interest_rate_spread  ...  credit_type  Credit_Score  \
0                   NaN  ...          EXP           758   
1                   NaN  ...

In [ ]:
print(df.isna().sum())

loan_limit                    3344
approv_in_adv                  908
loan_type                        0
loan_purpose                   134
Credit_Worthiness                0
open_credit                      0
business_or_commercial           0
loan_amount                      0
rate_of_interest             36439
Interest_rate_spread         36639
Upfront_charges              39642
term                            41
Neg_ammortization              121
interest_only                    0
lump_sum_payment                 0
property_value               15098
construction_type                0
occupancy_type                   0
Secured_by                       0
total_units                      0
income                        9150
credit_type                      0
Credit_Score                     0
co-applicant_credit_type         0
age                            200
submission_of_application      200
LTV                          15098
Region                           0
Security_Type       

In [ ]:
from sklearn.preprocessing import LabelEncoder

le = preprocessing.LabelEncoder()
df['loan_type'] = le.fit_transform(df['loan_type'])
print(df['loan_type'])

0         0
1         1
2         0
3         0
4         0
         ..
148665    0
148666    0
148667    0
148668    0
148669    0
Name: loan_type, Length: 148670, dtype: int64


In [ ]:
le = preprocessing.LabelEncoder()
df['open_credit'] = le.fit_transform(df['open_credit'])
print(df['open_credit'])

0         0
1         0
2         0
3         0
4         0
         ..
148665    0
148666    0
148667    0
148668    0
148669    0
Name: open_credit, Length: 148670, dtype: int64


In [ ]:
le = preprocessing.LabelEncoder()
df['occupancy_type'] = le.fit_transform(df['occupancy_type'])
print(df['occupancy_type'])

0         1
1         1
2         1
3         1
4         1
         ..
148665    1
148666    0
148667    1
148668    1
148669    1
Name: occupancy_type, Length: 148670, dtype: int64


In [ ]:
le = preprocessing.LabelEncoder()
df['business_or_commercial'] = le.fit_transform(df['business_or_commercial'])
print(df['business_or_commercial'])

0         1
1         0
2         1
3         1
4         1
         ..
148665    1
148666    1
148667    1
148668    1
148669    1
Name: business_or_commercial, Length: 148670, dtype: int64


In [ ]:

print(df['Status'])

0         1
1         1
2         0
3         0
4         0
         ..
148665    0
148666    0
148667    0
148668    0
148669    0
Name: Status, Length: 148670, dtype: int64


In [ ]:
from sklearn.model_selection import train_test_split

# Define the features and target variable
features = df[['open_credit', 'occupancy_type', 'loan_type', 'business_or_commercial']]
target = df['Status']

# Split the data into training and testing sets with a 80:20 ratio
X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

# Print the shape of the training and testing sets
print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)


Shape of X_train: (118936, 4)
Shape of X_test: (29734, 4)
Shape of y_train: (118936,)
Shape of y_test: (29734,)


In [ ]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report
knn = KNeighborsClassifier(n_neighbors=5)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)
print(classification_report(y_test, y_pred))


              precision    recall  f1-score   support

           0       0.76      1.00      0.86     22494
           1       0.00      0.00      0.00      7240

    accuracy                           0.76     29734
   macro avg       0.38      0.50      0.43     29734
weighted avg       0.57      0.76      0.65     29734



/usr/local/lib/python3.9/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.9/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.9/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


This program can predict with a 76% accuracy rate whether a applicant for a loan is likely to default on their loan. The program uses the data of applicants based on their open credit lines,occupancy, loan type, and whether the loan is business or commercial when providing a prediction. This program is useful when considering new applicants. Using this model we may be able to prevent accepting applicants who are more likely to default on the loan we provide. 